lib/bsdlib/bsd_os.c

```diff
@@ -47,12 +47,61 @@
 static const nrfx_uarte_t uarte_inst = NRFX_UARTE_INSTANCE(1);
 #endif
 
+#define NUM_SLEEPING_THREADS 10
+
 void IPC_IRQHandler(void);
 
+static inline int thread_id_hash(u32_t thread_id)
+{
+	/* TODO: A better idea on how to hash thread IDs without collision.
+	 * A linked list or an array to fill would introduce sync issues.
+	 */
+	return thread_id % NUM_SLEEPING_THREADS;
+}
+
+typedef struct {
+	u32_t thread_id;
+	bool sleeping;
+	struct k_sem sem;
+} sleeping_thread_t;
+
+sleeping_thread_t sleeping_threads[NUM_SLEEPING_THREADS];
+
+static void sleeping_threads_init(void)
+{
+	for (int i = 0; i < NUM_SLEEPING_THREADS; i++) {
+		k_sem_init(&sleeping_threads[i].sem, 0, 1);
+		sleeping_threads[i].thread_id = 0;
+		sleeping_threads[i].sleeping = false;
+	}
+}
+
+static void all_sleeping_threads_wake(void)
+{
+	for (int i = 0; i < NUM_SLEEPING_THREADS; i++) {
+		if (sleeping_threads[i].sleeping) {
+			k_sem_give(&sleeping_threads[i].sem);
+		}
+	}
+}
+
 int32_t bsd_os_timedwait(uint32_t context, uint32_t timeout)
 {
-	/* TODO: to be implemented */
-	return 0;
+	int converted_timeout = K_MSEC(timeout);
+
+	if (timeout == 0) {
+		converted_timeout = K_FOREVER;
+	}
+
+	int i = thread_id_hash(context);
+
+	sleeping_threads[i].thread_id = context;
+	sleeping_threads[i].sleeping  = true;
+
+	bool ret_val = k_sem_take(&sleeping_threads[i].sem, converted_timeout);
```

Comment:
the old PR had a discussion about using semphores for wakeup purpose and using array of threads indexed by hash

Short summary is that semaphores are not needed - you can sleem on `k_sleep` and wake up threads using `k_wakeup`. Due to this change `k_sleep` returns now the information if thread was woken up preemptively or if the sleep ended normally.
You can wakeup all threads sleeping by traversing list of threads e.g.
```
for (struct k_thread *thread = _kernel.threads; thread; thread = thread->next_thread)
   k_wakeup(thread);
```

Reply:
+1 for using `k_sleep` instead of semaphores. I think though we should keep track of which threads were blocked by `bsdlib` - waking up every single thread in the system sounds like an intrusive and error-prone solution.